### ISSUE TITLE
H2 idleTimeMs and GOAWAY frame

### ISSUE THREAD

**Issue by vadimi** [linkerd/linkerd#2021]
Issue Type:

- [x] Bug report
- [ ] Feature request

**What happened**:
I'm trying to setup `idelTimeMs` for `h2` router like that:
```
  client:
    clientSession:
      idleTimeMs: 5000
```
I have grpc service and client connected through linkerd `1.4.3`. After 5 seconds without any active RPCs I would expect lilnkerd to terminate that connection and send a `GOAWAY` frame to the client.

grpc client does **not** receive `GOAWAY` frame, but the service receives it.

**How to reproduce it (as minimally and precisely as possible)**:
Just configure idleTimeMs for h2 router.

**Anything else we need to know?**:

**Environment**:
- linkerd/namerd version, config files: linkerd 1.4.3
- Platform, version, and config files (Kubernetes, DC/OS, etc): running lilnkerd on local machine
- Cloud provider or hardware configuration:

**Comment by vadimi** [comment]
I tested `GOAWAY` sending a bit more and it looks like linkerd doesn't send it at all to clients. Here are a couple of tests I did:

- Gracefully terminated gRPC server which sends `GOAWAY` frame to all clients. Without linkerd it works fine, with linkerd clients never receive it.
- Configured `grpc.max_connection_idle_ms` on gRPC server side which terminates connection after some period of time and sends `GOAWAY` frame to all clients. Without linkerd all clients receive `GOAWAY` , with linkerd they don't.

Basically without `GOAWAY` frame clients don't know they got disconnected.

And another test - when I gracefully terminate linkerd it seems all clients receive `GOAWAY`.

**Comment by dadjeibaah** [comment]
Thanks for submitting this issue @vadimi and thanks for all the detailed tests you ran. We will take a look at this.

**Comment by dadjeibaah** [comment]
@vadimi after reproducing a scenario similar to what you described, it seems like things are working as intended AFAIK. I set up Linkerd to route requests from `nghttp` to an `HTTP/2` server that sent no frames to the `nghttp` client. I added a client configuration similar to use with an `idleTimeMs` of 5 seconds.

When firing a request to the "idle" HTTP/2 and waiting for about 6 seconds. Linkerd didn't send a `GOAWAY` frame to the `nghttp` client. I believe Linkerd, in this scenario, assumes the request is not satisfied, so the connections between the client <-> Linkerd and Linkerd <-> HTTP/2 server are still considered active. However, when terminating the client's request, Linkerd cached the connection to the HTTP/2 and terminated that connection after the configured 5 seconds. Let us know if you can reproduce this or if there is more information you may have on this issue.

**Comment by vadimi** [comment]
thanks for checking @dadjeibaah. The thing is that behavior is different from what gRPC expects - [https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md](https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md) (please see `MAX_CONNECTION_IDLE` section).

Basically if I route requests through linkerd my grpc clients don't know that grpc server sent them GOAWAY and they would never try to reconnect. This is useful in case I want clients to reconnect to linkerd cluster if I add one more node. Another example is if I have AWS network load balancer in front of linkerd - it has 350 seconds idle timeout after which it terminates connections. If I would configure max idle timeout in linkerd to let's say 300 seconds it will force grpc clients to reconnect and mitigate connection termination (since linkerd doesn't have any "keepalive" type of functionality).

If I run grpc client and server without linkerd in the middle all scenarios work as expected. I also compared linkerd behavior with envoy and envoy does send GOAWAY in all cases.

**Comment by vadimi** [comment]
Ok, thank you both for explanation.

**Comment by dadjeibaah** [comment]
@vadimi glad we could be of help! Let us know if you have other questions, in the meantime, I will go ahead and close this issue.
